tower-ready-cache/src/service.rs

```diff
@@ -0,0 +1,380 @@
+                    // replacement unready service has already been pushed.
+                }
+                Err(PendingError::Inner(key, e)) => {
+                    self.cancelations
+                        .swap_remove(&key)
+                        .expect("missing cancelation");
+                    return Err(error::Failed(key, e.into()));
+                }
+            }
+        }
+    }
+
+    /// Configures the ready service to be used by `call`.
+    ///
+    /// This may be called instead of invoking `Service::poll_ready`
+    pub fn poll_next_ready<Q: Hash + Equivalent<K>>(
+        &mut self,
+        key: &Q,
+    ) -> Poll<(), error::Failed<K>> {
+        match self.ready.get_full_mut(key) {
+            None => Ok(Async::NotReady),
+            Some((index, _, _)) => self.poll_next_ready_index(index),
+        }
+    }
+
+    /// Consume the `next_ready_index`.
+    pub fn take_next_ready_index(&mut self) -> Option<usize> {
+        self.next_ready_index.take()
+    }
+
+    /// Configures the ready service to be used by `call`.
+    ///
+    /// This may be called instead of `Service::poll_ready`.
+    ///
+    /// If the service is no longer ready and there are no equivalent unready
+    /// services, it is moved back into the ready set and `Async::NotReady` is
+    /// returned.
+    ///
+    /// If the service errors, it is removed and dropped and the error is returned.
+    ///
+    /// Otherwise, the `next_ready_index` is set with the provided index and
+    /// `Async::Ready` is returned.
+    pub fn poll_next_ready_index(&mut self, index: usize) -> Poll<(), error::Failed<K>> {
+        self.next_ready_index = None;
+
+        let (_, svc) = self.ready.get_index_mut(index).expect("index out of range");
+        match svc.poll_ready() {
+            Ok(Async::Ready(())) => {
+                self.next_ready_index = Some(index);
+                Ok(Async::Ready(()))
+            }
+            Ok(Async::NotReady) => {
+                // became unready; so move it back there.
+                let (key, svc) = self
+                    .ready
+                    .swap_remove_index(index)
+                    .expect("invalid ready index");
+
+                // If a new version of this service has been added to the
+                // unready set, don't overwrite it.
+                if !self.pending_contains(&key) {
+                    self.push_service(key, svc);
+                }
+
+                Ok(Async::NotReady)
+            }
+            Err(e) => {
+                // failed, so drop it.
+                let (key, _) = self
+                    .ready
+                    .swap_remove_index(index)
+                    .expect("invalid ready index");
+                Err(error::Failed(key, e.into()))
+            }
+        }
+    }
+
+    /// Calls the next ready service.
+    ///
+    /// `poll_next_ready_index` must have returned Ready before this is invoked;
```

Comment:
I _think_ it's okay to call any of the `poll_*ready*` methods before calling this, no? `poll_next_ready` for example.

Reply:
`poll_pending` is specifically _not_ sufficient though.